ddl/cache.lua

```diff
@@ -0,0 +1,120 @@
+--       parsed_func_name = <table> or <nil>
+--       -- a function ready to call
+--       callable = <function> or <nil>,
+--       -- string with an error: not nil only if setting callable fails
+--       error = <string> or <nil>,
+--     }
+--   },
+--   -- current schema version
+--   schema_version = <...>,
+-- }
+
+-- function returns nothing
+local function cache_build()
+    -- clear cache
+    cache.spaces = {}
+
+    if box.space._ddl_sharding_func == nil then
+        return
+    end
+
+    for _, tuple in box.space._ddl_sharding_func:pairs() do
+        local space_name = tuple[SPACE_NAME_IDX]
+        local func_name = tuple[SHARD_FUNC_NAME_IDX]
+        local func_body = tuple[SHARD_FUNC_BODY_IDX]
+
+        cache.spaces[space_name] = {
+            raw_tuple = tuple
+        }
+
+        if func_body ~= nil then
+            local sharding_func, err = loadstring('return ' .. func_body)
+            if sharding_func == nil then
+                cache.spaces[space_name].error =
+                    string.format("Body is incorrect in sharding_func for space (%s): %s",
+                    space_name, err)
+            else
+                cache.spaces[space_name].callable = sharding_func()
+            end
+        elseif func_name ~= nil then
+            -- we cannot save the function itself into the cache,
+            -- because the function can be changed in runtime and
+            -- there is no way to catch this change
+            local chunks = string.split(func_name, '.')
+            cache.spaces[space_name].parsed_func_name = chunks
+        end
+    end
+
+    cache.schema_version = box.internal.schema_version()
+
+end
+
+-- Rebuild cache if _ddl_sharding_func space changed.
+local function cache_set_trigger()
+    if box.space._ddl_sharding_func == nil then
+        return
+    end
+
+    local trigger_found = false
+
+    for _, func in pairs(box.space._ddl_sharding_func:on_replace()) do
+        if func == cache_build then
+            trigger_found = true
+            break
+        end
+    end
+
+    if not trigger_found then
+        box.space._ddl_sharding_func:on_replace(cache_build)
+    end
+end
+
+-- Get data from cache.
+-- Returns all cached data for "space_name" or nil.
+local function cache_get(space_name)
+    if space_name == nil then
+        return nil
+    end
+
+    -- using tarantool internal API.
+    -- this is not reliable, but it is the only way to track
+    -- schema_version changes. Fix it if a public method appears:
+    -- https://github.com/tarantool/tarantool/issues/6544
+    local schema_version = box.internal.schema_version()
+
+    if not cache then
+        cache = {}
+        box.atomic(cache_build)
+        cache_set_trigger()
+    end
+
+    -- rebuild cache if database schema changed
+    if schema_version ~= cache.schema_version then
+        box.atomic(cache_build)
+        cache_set_trigger()
+    end
+
+    return cache.spaces[space_name]
```

Comment:
Cited from the issue:

> Don't use the cache in a transaction? That's slow, but how else we can ensure that the cache will correspond to the state visible in the transaction?

It seems, current implementation does not reflect this point. So, if we're in a transaction, which can see a more fresh state than should be visible in the transaction. It could be achieved by using vinyl for `_ddl_sharding_func` (which has no sense) or by using MVCC mode for memtx (which is possible).

However I propose to clarify that the `ddl.bucket_id()` function is not transactional in the documentation instead of fixing it with performance loss. Because the performance loss may be a real user visible problem, but I can't imagine a practical case of a sharding function, which would depend on data in some space, which will be changed together with `_ddl_sharding_func` in a transaction as part of some migration. Really strage. Let's just leave a note in docs.

However I have one more practical question here. `box.atomic()` will raise an error inside a transaction, because tarantool does not support nested transactions:

```lua
tarantool> box.atomic(function() box.atomic(function() end) end)
---
- error: 'Operation is not permitted when there is an active transaction '
...
```

So, it the current implementation, `ddl.bucket_id()` cannot be used in a transaction at all. We shouldn't wrap the cache building into `box.atomic()` if we're in a transaction. There is [box.is_in_txn()](https://github.com/tarantool/tarantool/issues/3518) since 1.10.3 and 2.1.0, so I think that we can use it unconditionally.

Dropping `box.atomic()` here at all is okay too. I doubt that anyone ever will create `_ddl_sharding_func` in vinyl. A comment in the code (that clarifies our assumptiouns) is more than enough.

I vote for dropping `box.atomic()`. If you're agree, I expect the following changes:

* Drop `box.atomic()`.
* Add a comment into the code regarding our assumptions (`_ddl_sharding_func` is memtx space, so `:pairs()` on it is atomic — AFAIU even in MVCC mode).
* Add a comment into README: `ddl.bucket_id()` is not transactional in the sense that it catches up `_ddl_sharding_func` changes immediatelly: it may see changes that're not committed yet and may see a state from another transaction, which should not be visible in the current transaction.

----

I have a raw idea how to overcome the 'read uncommited' problem (we can use on_replace() + on_commit()), but we anyway can't have a separate cache version for a transaction, which changes a space we're interested in. I propose to don't do anything with it. It is interesting idea to implement in tarantool, though: I'll file an issue regarding this later.

Reply:
box.atomic() dropped, сomment added, bucket_id() descripton in REAMDE added

Reply:
Okay.

Filed https://github.com/tarantool/tarantool/discussions/6800 with a dump of my current thoughts.